File: inchcape-mvp/frontend/src/api/pipeline.ts
import { apiFetch } from './config';

// Coincide 1:1 con app/schemas/pipeline.py del backend.
export interface RecomendacionSKU {
  RANKING_SCORE: number;
  SKU: string;
  MARCA: string;
  FAMILIA: string;
  ABC: string;
  ROTACION_6M: number;
  N_PEDIDOS: number;
  N_LINEAS: number;
  CANT_TOTAL: number;
  VOLUMEN_M3: number;
  PESO_KG: number;
  ZONA_ACTUAL: string;
  ZONA_RECOMENDADA: string;
  TIEMPO_LAYOUT_ACTUAL: number;
  TIEMPO_NUEVO_MIN: number;
  COSTO_ACTUAL_MIN: number;
  COSTO_NUEVO_MIN: number;
  AHORRO_ESTIMADO_MIN: number;
  AHORRO_PORCENTAJE: number;
  SCORE_PRIORIDAD: number;
  MOVIMIENTO: 'MOVER' | 'MANTENER';
  JUSTIFICACION: string;
  CLUSTER_ML: number;
  PERFIL_ML: string;
  PRIORIDAD_CLUSTER_RANK: number;
  INDICE_IMPACTO_CLUSTER: number;
}

export interface Kpis {
  sku_analizados: number;
  sku_movidos: number;
  porcentaje_sku_movidos: number;
  max_movimientos_permitidos: number;
  tiempo_actual_min: number;
  tiempo_optimizado_min: number;
  ahorro_min: number;
  reduccion_porcentaje: number;
  productividad_actual_lineas_hh: number;
  productividad_optimizada_lineas_hh: number;
  tiempo_promedio_actual_min_pedido: number;
  tiempo_promedio_optimizado_min_pedido: number;
}

export interface DecisionRegla {
  sku: string;
  regla_id: string;
  motivo: string;
}

export interface MetricasML {
  mejor_k: number;
  silhouette: number;
  interpretacion_silhouette: string;
  variables_usadas: string[];
  perfil_clusters: Record<string, number | string>[];
}

export interface BanderasActivas {
  usar_incompatibilidad_geometrica: boolean;
  usar_triage: boolean;
  usar_payback_real: boolean;
  usar_fifo: boolean;
}

export interface RespuestaPipeline {
  recomendaciones: RecomendacionSKU[];
  kpis: Kpis;
  banderas_activas: BanderasActivas;
  camino_decision_reglas: DecisionRegla[];
  ml: MetricasML;
}

export interface PesosScore {
  ahorro: number;
  rotacion: number;
  abc: number;
  facilidad_movimiento: number;
}

export function ejecutarPipeline(
  pesos_score?: PesosScore,
  porcentaje_max_movimiento?: number,
): Promise<RespuestaPipeline> {
  return apiFetch<RespuestaPipeline>('/pipeline/ejecutar', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ pesos_score: pesos_score ?? null, porcentaje_max_movimiento: porcentaje_max_movimiento ?? null }),
  });
}

File: inchcape-mvp/backend/app/schemas/pipeline.py
from pydantic import BaseModel


class SolicitudPipeline(BaseModel):
    pesos_score: dict[str, float] | None = None
    porcentaje_max_movimiento: float | None = None


class RecomendacionSKU(BaseModel):
    RANKING_SCORE: int
    SKU: str
    MARCA: str
    FAMILIA: str
    ABC: str
    ROTACION_6M: float
    N_PEDIDOS: float
    N_LINEAS: float
    CANT_TOTAL: float
    VOLUMEN_M3: float
    PESO_KG: float
    ZONA_ACTUAL: str
    ZONA_RECOMENDADA: str
    TIEMPO_LAYOUT_ACTUAL: float
    TIEMPO_NUEVO_MIN: float
    COSTO_ACTUAL_MIN: float
    COSTO_NUEVO_MIN: float
    AHORRO_ESTIMADO_MIN: float
    AHORRO_PORCENTAJE: float
    SCORE_PRIORIDAD: float
    MOVIMIENTO: str
    JUSTIFICACION: str
    CLUSTER_ML: int
    PERFIL_ML: str
    PRIORIDAD_CLUSTER_RANK: int
    INDICE_IMPACTO_CLUSTER: float

    model_config = {"populate_by_name": True}


class Kpis(BaseModel):
    sku_analizados: int
    sku_movidos: int
    porcentaje_sku_movidos: float
    max_movimientos_permitidos: int
    tiempo_actual_min: float
    tiempo_optimizado_min: float
    ahorro_min: float
    reduccion_porcentaje: float
    productividad_actual_lineas_hh: float
    productividad_optimizada_lineas_hh: float
    tiempo_promedio_actual_min_pedido: float
    tiempo_promedio_optimizado_min_pedido: float


class DecisionRegla(BaseModel):
    sku: str
    regla_id: str
    motivo: str


class MetricasML(BaseModel):
    mejor_k: int
    silhouette: float
    interpretacion_silhouette: str
    variables_usadas: list[str]
    # dict, no un modelo rígido: las columnas son las medias por variable
    # de `variables_usadas`, que varían según qué atributos tengan
    # variación en el lote -- es justo el "perfil de centroide en
    # variables reales" que pide la explicabilidad (no una lista fija).
    perfil_clusters: list[dict]


class RespuestaPipeline(BaseModel):
    recomendaciones: list[RecomendacionSKU]
    kpis: Kpis
    banderas_activas: dict[str, bool]
    camino_decision_reglas: list[DecisionRegla]
    ml: MetricasML
